frontend: Add Advanced search

sniok opened the pull request:
Advanced search is a new page under "Cluster" sidebar item.
It allows users to search any resource using custom queries. 
API discovery was added to enable loading any resource. 

Features

### Main part
search bar, resource selector, settings, namespace selector.
By default no resources are selected (so we don't send a lot of requests by default and also encourage people to select only those resources that are needed to reduce load on kube-apiserver)

![image](https://github.com/user-attachments/assets/ae01726e-4f1d-4d26-b286-fd339506f0a4)

### Resource Picker

API discovery provides a full list of all resources in the cluster. The picker has a fuzzy search based on group name and resource name. 

![image](https://github.com/user-attachments/assets/8d401f0a-7cc2-4946-951c-9094f74b0596)
![image](https://github.com/user-attachments/assets/9cc7b8c1-a517-43d1-9cab-af5d493ccc7c)

### Settings

There are currently only two settings, limit for the number of items per resource (so if there are 100k Pods, only pods will be excluded). And refetch interval since we're not opening websocket connections for the lists due to amount of resources that you can select.
![image](https://github.com/user-attachments/assets/b40f6ce9-e35e-4d04-99f1-60868a7f0a12)

### Error handling

If resources fail to load we show a message, with a tooltip 

![image](https://github.com/user-attachments/assets/b17334ef-61d0-463f-ba9f-9bc2a9495985)

### Multi-cluster support

Handles API discovery across multiple clusters, adds a cluster column

![image](https://github.com/user-attachments/assets/e32bdd9d-4290-4ef1-91c0-79922bbd9192)

### View YAML action

Since we currently don't support displaying details of all the resources, I've added a YAML button that opens an readonly editor. Resources that don't have Details page won't have clickable name in the Name column

![image](https://github.com/user-attachments/assets/abba6be8-bbf0-406d-9e7c-7ee0850b1840)


### Performance 

The search is using deferred value hooks along with an interruptible search that can yield which results in a search that works even on 100k resources without any major slowdowns or hangs.
By default the maximum limit for items per resource is very moderate (10,000) but users can increase it 

https://github.com/user-attachments/assets/1dce2757-c292-4de2-9cce-7d7895efaf8b

If the amount of results is more than a 1000 then a warning message is shown

![image](https://github.com/user-attachments/assets/57de3f4f-4b4c-4e52-bd52-db429dd2e6e2)

### Search state in the query params

Search stores most of the settings as the query parameters so clicking on back button will preserve the state and you can also bookmark the page for a quick access to the resources. I'd love to add an ability to save the url as a favorite within the app, but we'll take a look at it when we develop Favorites feature and refactor the sidebar. 

![image](https://github.com/user-attachments/assets/6e718edf-021d-4c8a-a6ec-8d09ba9bad63)

### Shows example queries that also select resources

Allows users to see it action in one-click 
![image](https://github.com/user-attachments/assets/778bfff3-a794-4fa8-b672-4335d71bb6c7)


### Testing done

Tested on various queries and resources on local k3s cluster, remote AKS cluster, local kwok cluster with >100,000 resources.
a11y automated checks

joaquimrocha reviewed (changes requested):
Great job, Olek! The whole view feels very snappy and very useful!
About the PR:
- Left a nit about the spinner, but if the styling we have on the Loader component is not matching, ignore it.
- Also, while testing it, I found:
  - We should add a title to the page/section for consistency
  - While resources are loading, the table is displayed but it's empty, so it says _No resources found_. If the search takes a couple of seconds, even with the spinner it's a bit confusing. So I think we should hide the table while searching.
  - Can we add a way to go to the advanced search from the regular search? Even if it's just a button in it which when clicked takes the user to the Advanced search with the string we had in the top search view?

sniok commented on frontend/src/components/advancedSearch/AdvancedSearch.tsx:
```diff
@@ -0,0 +1,149 @@
+  );
+
+  if (isLoading) {
+    return <CircularProgress />;
```
done

sniok commented:
> * Also, while testing it, I found:
>   
>   * We should add a title to the page/section for consistency

Agree, added a title

>   * While resources are loading, the table is displayed but it's empty, so it says _No resources found_. If the search takes a couple of seconds, even with the spinner it's a bit confusing. So I think we should hide the table while searching.

Added some additional logic to not display table when resources are still loading although there are still states when table briefly shows "no resources found" due to deferred query and results. For now I think it's okay but I have some ideas to improve this. 

>   * Can we add a way to go to the advanced search from the regular search? Even if it's just a button in it which when clicked takes the user to the Advanced search with the string we had in the top search view?

Added an entry to the global search

![image](https://github.com/user-attachments/assets/0941a38e-b2e1-4e87-90be-4a05bfe9d547)

illume commented:
@joaquimrocha @sniok can you please let me know what testing you've done in the PR description?

illume commented:
Was an issue written for this? Can you please link it into the PR description?

illume commented:
This gives a better idea of the number of tests (from make frontend-test):
![image](https://github.com/user-attachments/assets/38b63a79-4ba1-4fab-9bb0-96307868ae96)

There's no e2e tests.

illume commented:
I wanted to check these topics:

- [ ] does it search CRs?
- [x] does it search kubernetes labels (popular request)
- [ ] server load (an idea of how many requests and how much data for a search)
- [ ] Should we put beta/alpha on search over all?

This was the comment from before about searching kubernetes labels.

> labels are important to users (people use them with kubectl a lot, and it's one of the top requested dashboard features). https://github.com/kubernetes-sigs/headlamp/issues/2547 Does your search already look at labels?
>> in the advanced search you can use any property from the resources to search by, for example if you want to use labels you can type the query "metadata.labels.env === 'prop'". You can put any js expression on full resource yaml/json. It supports selecting any kind of resource that exists in the cluster so it supports custom resources


Should we be putting beta on search over all? If we have things we still want to develop, then alpha is the more appropriate label. I know we still want to do backend caching, and searching with llms. So my suggestion is that we put the alpha label on search overall (or beta, but beta implies that we are functionally complete but still want to polish things).

illume commented:
Would you mind breaking up the two largest commits if possible? So it's easier to review.

| Commit      | Message                                             | +Lines | -Lines |
|-------------|------------------------------------------------------|--------|--------|
| 5a6b56fdc  | frontend: GlobalSearch: Add entry for Advanced Searc |     22 |      2 |
| fbdf20a96  | frontend: Add Advanced Search entry to sidebar       |     94 |      0 |
| 665dd70f7  | frontend: Add route for Advanced Search component    |     11 |      0 |
| b7c6b4e9a  | frontend: Update Route interface to support ExoticCo |      2 |      2 |
| 93782f470  | frontend: Add AdvancedSearch component               |   1892 |     44 |
| e9d0c2392  | frontend: DetailsDrawer: Remove unnecessary url chan |      0 |     13 |
| b0d03ec28  | frontend: Update security kindGroups to include Clus |      1 |      1 |
| 7f4200a40  | frontend: Implement API resource discovery with aggr |    867 |      0 |

sniok commented on frontend/src/components/advancedSearch/utils/searchWithQuery.tsx:
```diff
@@ -0,0 +1,74 @@
+  let queryFn: any;
+  const functionBody = query;
+  try {
+    queryFn = eval(`({ ${topLevelKeys.join(', ')} }) => { return ${functionBody}; }`);
```
there's no security risk here, this is client side

illume replied:
Opening this again. Please don't close this until I mark it as resolved.

joaquimrocha reviewed (changes requested):
Left a comment about an untranslatable string. There may be possibly others. Also, I feel it should have a "clear" button in the search entry to reset the search, as selecting one of the suggested searches is nice but going back to the previous state involves clicking the side bar again and that's not a great UX.

sniok commented:
Added a clear button and updated some missing translation strings

illume reviewed (changes requested):
Very cool. 

Added some notes below...

- [ ] Can you please add a kubernetes label search example in there? ( I think you mentioned something like `metadata.labels.env === 'prop'"` ? ). This is a highly requested feature.
- [ ] Would you mind breaking up the two largest commits if possible? So it's easier to review. The AdvancedSearch commit (approx 2000 line diff) for example has multiple components that can be broken up into separate commits.
- [ ] Can you please add some more tests? I think it's up to about 46% of functions added in this PR are tested.

![Screenshot 2025-06-18 150935](https://github.com/user-attachments/assets/1bbd9cf6-3e88-44dc-870c-81715d810e68)


- [ ] can you both please let me know what testing you've done in the PR description? I'm especially interested in:
    - [ ] a11y 
    - [ ] load (lots of resource types) Here's a not too uncommon setup... Crossplane (50–300+), Flux (10–30), Gateway API (10–20), Cluster API (50–150+), Cert-Manager (10–20), Prometheus Operator (10–20), Istio/Linkerd (20–50+). Openshift and Crossplane can have over 1000+ each if they're fully configured. Total resource types: CRDs 150–600+ to 1000s at the high end. I'm concernted a search could do 1000s of requests in this situation, but didn't check if there are limits for this.
    - [ ] user testing with people. I'm concerned about users liking the JS/TypeScript based search. Especially since many of our users don't know JS (or know it well). The examples are nice, and the AI would improve things. I'd still be interested to know how a few people act when presented with 6 questions (not in the examples) if they could figure it out. Would they like it? I think this could be done before or after merge. I feel we can add quite a few examples as people ask how to search for things.



- [ ] Regarding the security issue, Did you think about parsing the input somehow? (maybe https://github.com/EricSmekens/jsep ?) So it's a more limited set of input allowed? What about confirming with the user somehow before running it when coming in from a URL? Screenshot of placing alert() in the URL below.

![image](https://github.com/user-attachments/assets/616315b2-f417-46ad-a861-56400b2a9781)

- [ ] Did you consider having fuzzy / free text search being done if there are no results? Like if I type results in the top search, they should be available in the advanced search (assuming no results)? Did you consider how these two types of search be combined into one?
- [ ] when I enter a search, a result comes up, and then press enter again it gets stuck
![image](https://github.com/user-attachments/assets/a33e0195-48cb-45a6-80d0-fddb8389686d)

sniok commented:
I thought about security of eval'ing the expression a bit and I will remove the ability to load the expression from query parameter. With that change the only code that will be executed in the search would the one user typed.

sniok commented:
thanks for taking a look @joaquimrocha 

> I tested again and noticed that even though you mentioned you made it no longer show the "No data" while loading, I think it's always showing this when I choose one of the predefined examples.

Updated this, should behave better now

> More importantly, I tried typing "!!data" in the advanced search and can never activate this (i.e. pressing enter or clicking is not adding ti ) IFF we are already in the advanced search view. 

Fixed

> Also, when we are not in the adv search view and it does work to go from global search, it always tries to search for a metadata.name equal to what I typed. If I typed something like "metadata.name.startsWith('kube')", it should just execute that search with the core resources selected or similar. Maybe a heuristic is: if string is not a valid name (we validate this for e.g. namespace names), we should just run the string as a query.

I actually don't think we should do this. Global search in the top bar uses fuzzy search by name, advanced search uses queries. I think it will be easier for users to understand this separation.

![image](https://github.com/user-attachments/assets/4d1d3e4d-2ca8-4fcb-b68d-c75bc5454ffe)

We explicitly tell users that this will search by name. So when user is linked to the Advanced Search it automatically replaced by an almost equivalent expression `metadata.name === abc` which also introduces people to how queries work. Adding a hidden logic for cases where people entered expression into the Global search would only be somewhat beneficial for people who are already familiar with advanced search and writing queries but will be confusing for new users. For example if a new user added a dot at the end by accident then we should still show `metadata.name ===`

> The enter key seems to still input an newline in the text entry and I understood from your comments that this was no longer the case.

Enter key always made a new line, just before it wasn't expanding the input, but now it expands input properly

joaquimrocha commented:
/lgtm
/approve
